### PR TITLE
MaskRCNN support PR

### PR DESCRIPTION
**Pull request by SundarRajan28**
- Added support in meta data for reading masks
- Added bool variables in functions related to COCOreader to switch reading masks on and off
- Added support for resize_mirror_normalize RPP function and added nodes for calling the respective augmentation
- Added a coco segmentation example python file to run and check the segmentation build

### CODE REVIEW

File: rocAL/rocAL/include/image_loader.h [hunk 1]
```diff
@@ -83,7 +83,7 @@ class ImageLoader : public LoaderModule {
     size_t _prefetch_queue_depth; // Used for circular buffer's internal buffer
     size_t _image_counter = 0;//!< How many images have been loaded already
     size_t _remaining_image_count;//!< How many images are there yet to be loaded
-    bool _decoder_keep_original = false;
+    bool _decoder_keep_original = true;
```
Comment (swetha097) [inline rocAL/rocAL/include/image_loader.h:86]: Any specific reason we changed the default value of _decoder_keep_original to true?

File: rocAL/rocAL/source/bounding_box_graph.cpp [hunk 2]
```diff
@@ -20,12 +20,101 @@ OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
...
+            }
+        }
+
+        int m = 0;
+        for (uint j = 0; j < bb_count; j++)
+        {
+            BoundingBoxCord box;
+            box.l = coords_buf[m++];
+            box.t = coords_buf[m++];
+            box.r = coords_buf[m++];
```
Comment (swetha097) [inline rocAL/rocAL/source/bounding_box_graph.cpp:100]: Declare coords_buf as BoundingBoxCords type & directly push to bb_coords instead of assigning to box , then pushing it

Refer meta_node_flip.cpp for reference

File: rocAL/rocAL/source/coco_meta_data_reader.cpp [hunk 3]
```diff
@@ -237,27 +304,39 @@ void COCOMetaDataReader::read_all(const std::string &path)
...
+                }
             }
         }
         else
         {
             parser.SkipValue();
         }
     }
-    for (auto &elem : _map_content)
+    /*for (auto &elem : _map_content)
```
Comment (swetha097) [inline rocAL/rocAL/source/coco_meta_data_reader.cpp:339]: Why was this part commented out?
This part is necessary for SSD pipeline to convert label_id to a continous_label_id (inv_map)

Comment (SundarRajan28) [reply rocAL/rocAL/source/coco_meta_data_reader.cpp]: Resolved in latest commit

File: rocAL/rocAL/include/image_loader.h (hunk 1 above)
Comment (SundarRajan28) [reply rocAL/rocAL/include/image_loader.h]: Resolved in latest commit

File: rocAL/rocAL/source/bounding_box_graph.cpp [hunk 4]
```diff
@@ -20,12 +20,101 @@ OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
...
+                    index = 1;
+                }
+            }
+        }
+
+        int m = 0;
+        for (uint j = 0; j < bb_count; j++)
+        {
+            BoundingBoxCord box;
+            box.l = coords_buf[m++];
```
[the remark this answers is not in the document]
Comment (SundarRajan28) [reply rocAL/rocAL/source/bounding_box_graph.cpp]: Resolved in latest commit

File: rocAL/rocAL_pybind/example/coco_pipeline_segmentation.py [hunk 5]
```diff
@@ -0,0 +1,587 @@
...
+            self.bb_2d_numpy[1] *= self.img_roi_size2d_numpy_wh[1]
+            self.bb_2d_numpy[2] *= self.img_roi_size2d_numpy_wh[0]
+            self.bb_2d_numpy[3] *= self.img_roi_size2d_numpy_wh[1]
+
+            print("\nAfter : self.bb_2d_numpy\n", self.bb_2d_numpy)
+
+            self.bb_2d_numpy = np.reshape(self.bb_2d_numpy, (-1, 4)).tolist()
+
+            val_dataset = get_val_dataset()
+            inv_map = {v: k for k, v in val_dataset.items()}
```
[the remark this answers is not in the document]
Comment (SundarRajan28) [reply rocAL/rocAL_pybind/example/coco_pipeline_segmentation.py]: Resolved in latest commit

File: rocAL/rocAL_pybind/example/coco_pipeline_segmentation.py [hunk 6]
```diff
@@ -0,0 +1,587 @@
...
+
+            self.inv_label = []
+            for element in self.label_2d_numpy:
+                self.inv_label.append(inv_map[element])
```
[the remark this answers is not in the document]
Comment (SundarRajan28) [reply rocAL/rocAL_pybind/example/coco_pipeline_segmentation.py]: Resolved in latest commit

File: utilities/rali/rali_unittests/rali_unittests.cpp [hunk 7]
```diff
@@ -160,7 +160,7 @@ int test(int test_case, int reader_type, int pipeline_type, const char *path, co
                 std::cout << "\n json_path has to be set in rali_unit test manually";
                 exit(0);
             }
-            meta_data = raliCreateCOCOReader(handle, json_path, true);
+            meta_data = raliCreateCOCOReader(handle, json_path, true, true);
```
Comment (swetha097) [inline utilities/rali/rali_unittests/rali_unittests.cpp:163]: Please have a mask variable to pass for either segmentation / detection instead of hard-coding to true

File: rocAL/rocAL/source/rali_api_augmentation.cpp [hunk 8]
```diff
@@ -1699,7 +1699,7 @@ RALI_API_CALL raliCropMirrorNormalize(RaliContext p_context, RaliImage p_input,
         context->capture_error(e.what());
         ERR(e.what())
     }
-    return input; // Changed to input----------------IMPORTANT
+    return output; // Changed to input----------------IMPORTANT
```
Comment (swetha097) [inline rocAL/rocAL/source/rali_api_augmentation.cpp:1702]: Remove the comment

File: rocAL/rocAL/source/rali_api_augmentation.cpp [hunk 9]
```diff
@@ -541,6 +543,56 @@ raliResize(
...
+        rmn_node->init(0, 1, mirror);
+        if (context->master_graph->meta_data_graph())
+            context->master_graph->meta_add_node<ResizeMirrorNormalizeMetaNode,ResizeMirrorNormalizeNode>(rmn_node);
+    }
+    catch(const std::exception& e)
+    {
+        context->capture_error(e.what());
+        ERR(e.what())
+    }
+    return input; // Changed to input
```
[the remark this answers is not in the document]
Comment (SundarRajan28) [reply rocAL/rocAL/source/rali_api_augmentation.cpp]: Resolved in the latest commit. Modified the same for crop_mirror_normalize as well.

File: rocAL/rocAL_pybind/amd/rali/pipeline.py [hunk 10]
```diff
@@ -128,14 +128,17 @@ def store_values(self, operator):
         if(operator.data in self._check_ops):
             self._tensor_layout = operator._output_layout
             self._tensor_dtype = operator._output_dtype
-            self._multiplier = list(map(lambda x: 1/x ,operator._std))
-            self._offset = list(map(lambda x,y: -(x/y), operator._mean, operator._std))
-            #changing operator std and mean to (1,0) to make sure there is no double normalization
-            operator._std = [1.0]
```
[the remark this answers is not in the document]
Comment (SundarRajan28) [reply rocAL/rocAL_pybind/amd/rali/pipeline.py:134]: Resolved in the latest commit.

File: rocAL/rocAL_pybind/rali_pybind.cpp [hunk 11]
```diff
@@ -157,6 +157,27 @@ namespace rali{
         return py::cast<py::none>(Py_None);
     }
 
+    py::object wrapper_Mask_count(RaliContext context, py::array_t<int> array)
+    {
+        auto buf = array.request();
+        int* ptr = (int*) buf.ptr;
+        // call pure C++ function
+        int count = raliGetMaskCount(context,ptr);
+
```
[the remark this answers is not in the document]
Comment (SundarRajan28) [reply rocAL/rocAL_pybind/rali_pybind.cpp]: Resolved in the latest commit

File: rocAL/rocAL_pybind/example/coco_pipeline_segmentation.py [hunk 12]
```diff
@@ -0,0 +1,587 @@
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+import torch
+import ctypes
```
[the remark this answers is not in the document]
Comment (SundarRajan28) [reply rocAL/rocAL_pybind/example/coco_pipeline_segmentation.py]: Resolved in the latest commit

File: rocAL/rocAL/source/meta_node_resize_mirror_normalize.cpp [hunk 13]
```diff
@@ -0,0 +1,150 @@
...
+            {
+                float l = 1 - box.r;
+                box.r = 1 - box.l;
+                box.l = l;     
+            }
+
+            bb_coords.push_back(box);
+            bb_labels.push_back(labels_buf[j]);
+        }
+        if (bb_coords.size() == 0)
```
[the remark this answers is not in the document]
Comment (SundarRajan28) [reply rocAL/rocAL/source/meta_node_resize_mirror_normalize.cpp]: Resolved in the latest commit.

File: utilities/rali/rali_unittests/rali_unittests.cpp (hunk 7 above)
Comment (SundarRajan28) [reply utilities/rali/rali_unittests/rali_unittests.cpp]: Resolved in the latest commit. I have added a seperate pipeline type for the user to choose and run the coco segmentation example in the unittests. I also added a test case for resize_mirror_normalize augmentation as well.

File: amd_openvx_extensions/amd_rpp/source/ResizeMirrorNormalizePD.cpp [hunk 14]
```diff
@@ -0,0 +1,256 @@
...
+	RppiSize *srcDimensions;
+	RppiSize maxSrcDimensions;
+	RppiSize *dstDimensions;
+	RppiSize maxDstDimensions;
+	RppPtr_t pSrc;
+    RppPtr_t pDst;
+	vx_float32  *mean;
+    vx_float32 *std_dev;
+    vx_uint32 *mirror;
+    vx_uint32 chnShift; //NHWC to NCHW
```
[the remark this answers is not in the document]
Comment (SundarRajan28) [reply amd_openvx_extensions/amd_rpp/source/ResizeMirrorNormalizePD.cpp]: Resolved in the latest commit.

File: rocAL/rocAL/include/master_graph.h [hunk 15]
```diff
@@ -78,8 +80,9 @@ class MasterGraph
     Image *create_image(const ImageInfo &info, bool is_output);
     Image *create_loader_output_image(const ImageInfo &info);
     MetaDataBatch *create_label_reader(const char *source_path, MetaDataReaderType reader_type);
+    MetaDataBatch *create_coco_meta_data_reader(const char *source_path, bool is_output, bool mask);
```
[the remark this answers is not in the document]
Comment (SundarRajan28) [reply rocAL/rocAL/include/master_graph.h]: Resolved in the latest commit. Removed the first function with 3 arguments in favour of the function with 8 arguments.

File: rocAL/rocAL/source/rali_api_meta_data.cpp [hunk 16]
```diff
@@ -328,6 +329,62 @@ RALI_API_CALL raliGetBoundingBoxCords(RaliContext p_context, float* buf)
     }
 }
 
+unsigned
+RALI_API_CALL raliGetMaskCount(RaliContext p_context, int* buf)
+{
+    if (!p_context)
```
[the remark this answers is not in the document]
Comment (SundarRajan28) [reply rocAL/rocAL/source/rali_api_meta_data.cpp]: Resolved in the latest commit.

File: rocAL/rocAL/source/rali_api_meta_data.cpp [hunk 17]
```diff
@@ -328,6 +329,62 @@ RALI_API_CALL raliGetBoundingBoxCords(RaliContext p_context, float* buf)
...
+            size += polygon_count;
+        }
+    }
+    return size;
+}
+
+void
+RALI_API_CALL raliGetMaskCoordinates(RaliContext p_context, int *bufcount, float *buf)
+{
+    if (!p_context)
```
[the remark this answers is not in the document]
Comment (SundarRajan28) [reply rocAL/rocAL/source/rali_api_meta_data.cpp]: Resolved in the latest commit.

File: rocAL/rocAL/include/meta_data.h [hunk 18]
```diff
@@ -152,6 +165,7 @@ struct MetaDataBatch
     virtual void clear() = 0;
     virtual void resize(int batch_size) = 0;
     virtual int size() = 0;
+    virtual int mask_size() = 0;
```
[the remark this answers is not in the document]
Comment (SundarRajan28) [reply rocAL/rocAL/include/meta_data.h]: Resolved in the latest commit.

File: rocAL/rocAL/include/meta_data.h [hunk 19]
```diff
@@ -193,6 +209,10 @@ struct LabelBatch : public MetaDataBatch
     {
         return _label_id.size();
     }
+    int mask_size() override
+    {
+        return 0;
+    }
     std::shared_ptr<MetaDataBatch> clone() override
```
[the remark this answers is not in the document]
Comment (SundarRajan28) [reply rocAL/rocAL/include/meta_data.h]: Resolved in the latest commit.

File: rocAL/rocAL/include/meta_node_resize_mirror_normalize.h [hunk 20]
```diff
@@ -0,0 +1,42 @@
+/*
+Copyright (c) 2019 - 2020 Advanced Micro Devices, Inc. All rights reserved.
```
[the remark this answers is not in the document]
Comment (SundarRajan28) [reply rocAL/rocAL/include/meta_node_resize_mirror_normalize.h]: Resolved in the latest commit.

File: rocAL/rocAL/include/rali_api_meta_data.h [hunk 21]
```diff
@@ -125,6 +125,13 @@ extern "C" void RALI_API_CALL raliGetImageLabels(RaliContext rali_context, int*
 /// \return The size of the buffer needs to be provided by user to get bounding box info for all images in the output batch.
 extern "C" unsigned RALI_API_CALL raliGetBoundingBoxCount(RaliContext rali_context, int* buf);
 
+///
+/// \param rali_context
+/// \param image_idx the imageIdx in the output batch
+/// \return The size of the buffer needs to be provided by user to get mask box info associated with image_idx in the output batch.
+extern "C" unsigned RALI_API_CALL raliGetMaskCount(RaliContext rali_context, int* buf );
+extern "C" void RALI_API_CALL raliGetMaskCoordinates(RaliContext rali_context, int* bufcount, float* buf);
```
[the remark this answers is not in the document]
Comment (SundarRajan28) [reply rocAL/rocAL/include/rali_api_meta_data.h:138]: Resolved in the latest commit. Added comments explaining the usage of the two mask functions.

File: rocAL/rocAL/source/bounding_box_graph.cpp [hunk 22]
```diff
@@ -20,12 +20,97 @@ OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
...
+void BoundingBoxGraph::process(MetaDataBatch *meta_data, bool segmentation)
 {
     for (auto &meta_node : _meta_nodes)
     {
-        meta_node->update_parameters(meta_data);
+        meta_node->update_parameters(meta_data, segmentation);
+    }
+}
+
+//not required since the bbox are normalized in the very beggining -> remove the call in master graph also
```
[the remark this answers is not in the document]
Comment (SundarRajan28) [reply rocAL/rocAL/source/bounding_box_graph.cpp]: Resolved in the latest commit. This function is called for both mask and SSD pipelines but updates the mask meta_data depending on the value of segmentation bool variable

File: rocAL/rocAL/source/master_graph.cpp [hunk 23]
```diff
@@ -400,6 +401,24 @@ MasterGraph::sequence_frame_timestamps(std::vector<std::vector<float>> &sequence
     _sequence_frame_timestamps_vec.pop_back();
 }
 
+std::vector<uint32_t>
```
[the remark this answers is not in the document]
Comment (SundarRajan28) [reply rocAL/rocAL/source/master_graph.cpp:404]: The returned vector is popped from the back of _resize_width and stored in a local variable. So I don't think it can be passed as a reference. Leaving it as it is in the latest commit.

File: rocAL/rocAL/source/rali_api_augmentation.cpp [hunk 24]
```diff
@@ -541,6 +543,56 @@ raliResize(
...
+        // For the resize mirror normalize resize node, user can create an image with a different width and height
+        ImageInfo output_info = input->info();
+        output_info.width(dest_width);
+        output_info.height(dest_height);
+        output = context->master_graph->create_image(output_info, is_output);
+        // For the nodes that user provides the output size the dimension of all the images after this node will be fixed and equal to that size
+        output->reset_image_roi();
+
+        std::shared_ptr<ResizeMirrorNormalizeNode> rmn_node =  context->master_graph->add_node<ResizeMirrorNormalizeNode>({input}, {output});
+        //rmn_node->init(mean_acutal, std_actual, mirror);
```
[the remark this answers is not in the document]
Comment (SundarRajan28) [reply rocAL/rocAL/source/rali_api_augmentation.cpp:583]: Talked with @swetha097 and the normalization doesn't happen in this node. In python end, we save the values of mean, std in the pipeline
 and they are passed to CopyToTensor for normalization. So this init function call is correct.

**Comment by SundarRajan28** [comment]
Closing the PR due to high number of changed files.
